### ISSUE TITLE
Create intermediate files in tmp dir with unique names to allow building several targets in parallel

### ISSUE THREAD

**Issue by mrdudz** [cc65/cc65#2446]
### Discussed in https://github.com/cc65/cc65/discussions/2445

<div type='discussions-op-text'>

<sup>Originally posted by **Fabrizio-Caruso** May  8, 2024</sup>
CC65 seems to make it harder or impossible to build the same game for different targets or different games when these share some common source files.
I don't have the same problem with gcc and Z88DK, which both make parallel builds work with no issue.
I suppose the problem is that CC65 generates some temporary .o files on the same directory as the .c, which it also deletes. This creates race conditions on such files:
```
ld65: Error: Input file 'cross_lib/display/display_macros.o' not found
ld65: Error: Input file 'cross_lib/display/display_macros.o' not found
ld65: Error: Input file 'cross_lib/display/display_macros.o' not found
Fatal error: Cannot open input file 'cross_lib/sound/cc65/vic20/vic20_sounds.s': No such file or directory
ld65: Error: Input file 'cross_lib/display/display_macros.o' not found
ld65: Error: Input file 'cross_lib/display/display_macros.o' not found
make: *** [Makefile_common:628: vic20_exp_16k] Error 1
make: *** [Makefile_common:628: vic20_exp_16k] Error 1
make: *** [Makefile_common:628: vic20_exp_16k] Error 1
make: *** [Makefile_common:628: vic20_exp_16k] Error 1
make: *** [Makefile_common:628: vic20_exp_16k] Error 1
make: *** [Makefile_common:628: vic20_exp_16k] Error 1
ld65: Error: Input file 'cross_lib/display/display_macros.o' not found
make: *** [games/shoot/makefiles/Makefile.override:9: vic20_exp_16k] Error 1
```
Is there a way to tell CC65 to avoid this? Either by putting such .o in memory or in a separate temporary folder per target or by using some sort of locks or any other parallel-friendly mechanism?

If this is not possible yet, this would be a nice feature to add in the future. 

I asked the Z88DK devs how they do it. They say that when I run their compiler with several source files in one line, no .o is generated.</div>

**Comment by GorillaSapiens** [comment]
isn't this really a build environment issue?  if i "gcc -c -O2 foo.c" in parallel with a "gcc -c -g foo.c", i don't get two different foo.o files generated with different options.  whichever one finishes last will just clobber what came before it.

**Comment by GorillaSapiens** [comment]
isn't this really a build environment issue? if i "gcc -c -O2 foo.c" in parallel with a "gcc -c -g foo.c", i don't get two different foo.o files generated with different options. whichever one finishes last will just clobber what came before it.

if i want two parallel builds of the same code with different options, i set up my build Makefiles or whatever to work in two different build directories, so things don't clobber each other.

**Comment by GorillaSapiens** [comment]
i don't see any unexpected extra files being generated here.

generation and subsequent deletion of intermediate files foo.s and foo.o are expected.
i would think gcc would do pretty much the same thing.
i certinaly wouldn't expect two instances of gcc to work well on the same file in the same working directory.

<pre>
gorilla@guest:~/tmp$ ls
foo.c
gorilla@guest:~/tmp$ cat foo.c
#include <stdio.h>

void main(void) {
printf("hello, world\n");
}
gorilla@guest:~/tmp$ strace ~/cc65/bin/cl65 foo.c >& out.txt
gorilla@guest:~/tmp$ ls
foo  foo.c  out.txt
gorilla@guest:~/tmp$ grep unlink out.txt 
unlink("foo.s")                         = 0
unlink("foo.o")                         = 0
gorilla@guest:~/tmp$ 
</pre>

**Comment by GorillaSapiens** [comment]
nevermind.  i get it now.

<pre>
gorilla@guest:~/tmp$ strace gcc foo.c |& grep unlink
unlink("/tmp/ccJPEe6b.res")             = 0
unlink("/tmp/ccGlMoGl.o")               = 0
unlink("/tmp/cchwrCiy.s")               = 0

gorilla@guest:~/tmp/bae$ strace ~/cc65/bin/cl65 ../foo.c |& grep unlink
unlink("../foo.s")                      = 0
unlink("../foo.o")                      = 0
</pre>

**Comment by rofl0r** [comment]
some observations:
- for make's target rules to function properly, mere existance of a file already indicates "the file exists and is ready to use"
- this means an .o or .s file that's still being written to will be detected as ready and used in subsequent actions in a parallel build when it's not completely written yet, leading to bugs.
- therefore, any output file should first be a tempfile, and when it's completely written, rename()'d to the final filename - this is atomic (at least on unix)
- however - at least on UNIX - rename() doesn't work across different filesystems - /tmp/ may reside on a different disk or even in ram (tmpfs). doing a rename() in such a case will lead to ENXIO.
- that means the tempfile should be created in the same dir the target .s/.o will be moved to, but not in /tmp.

**Comment by rofl0r** [comment]
> Also, make correctly waits for each prerequisite file to be fully built before using it in dependent rules — even when running with multiple threads via make -jN.

you're right - if there's a rule to build a file that's needed as a prereq for another rule, then the file will be marked as ready for next steps only when the rule that causes its creation completes and the file exists afterwards. so in a non-recursive make invocation, this should indeed not be an issue so the rename() isn't strictly necessary. IIRC i encountered the issue i described when yacc produced two files, but i didnt correctly document this fact, but had the generated header as a prereq for another file - and there it was used while it was still being written to. 

> Only intermediate files have to be temp files, not every file

right. if gcc (or any other compiler) creates an intermediate .s file from foo.c to assemble the target (foo.o), then it doesn't matter where the temporary .s file is saved - only that it has a unique name.

**Comment by mrdudz** [comment]
fixed by #2660
